### PR TITLE
Eliminate bd2k dependencies.  Partially addresses #1936.

### PR DESCRIPTION
**Pull request by DailyDreaming**
Partially addresses #1936.

Eliminates the `bd2k.util.files` dependency (for py2/3 compatibility).

Should be a part of #1937, but I'm not sure what's happening with that PR and also a wholesale import of the bd2k lib might be more than we need?  A lot of bd2k seems unnecessary to import.

### CODE REVIEW

File: src/toil/common.py [hunk 1]
```diff
@@ -233,9 +232,7 @@ def parseIntList(s):
         setOption("scaleInterval", float)
         setOption("metrics")
         setOption("preemptableCompensation", float)
-        require(0.0 <= self.preemptableCompensation <= 1.0,
-                '--preemptableCompensation (%f) must be >= 0.0 and <= 1.0',
-                self.preemptableCompensation)
+        assert 0.0 <= self.preemptableCompensation <= 1.0, '--preemptableCompensation (%f) must be >= 0.0 and <= 1.0' % self.preemptableCompensation
```
Comment (diekhans) [inline src/toil/common.py]: This looks like an user error and hence should raise an exception, not be an assert.

File: src/toil/job.py [hunk 2]
```diff
@@ -708,8 +707,7 @@ def defer(self, function, *args, **kwargs):
 
         :param dict kwargs: The keyword arguments to the function
         """
-        require(self._fileStore is not None, 'A deferred function may only be registered with a '
-                                             'job while that job is running.')
+        assert self._fileStore is not None, 'A deferred function may only be registered with a job while that job is running.'
         self._fileStore._registerDeferredFunction(DeferredFunction.create(function, *args, **kwargs))
```
Comment (diekhans) [inline src/toil/job.py]: also should raise and exception, not an assert

**Comment by diekhans** [comment]
losing the bd2k lib is a blocking change for PY3 compatibility.

bd2k lib is very hard to port to PY3 due to using doctest as a testing framework.  Since it compares printout for humans, not object, and can look slightly different in PY2 vs PY3.  The only way to have doctest work is to have two copies of the function doc.  It can't be dynamically built, as the doc strings are defined at parse time, not execution type.

**Comment by diekhans** [comment]
p.s. there appears to be a way in the doctest class to save the tests as unittests.  This will then be easy to convert to pytest.

**Comment by DailyDreaming** [comment]
That would be useful for some of the rest of bd2k that has testing that would be good to retain.  I wouldn't mind merging this as is if you approve and doing the rest in chunks when I have time.

**Comment by DailyDreaming** [comment]
Flaky test:
- src.toil.test.provisioners.aws.awsProvisionerTest.AWSAutoscaleTestMultipleNodeTypes.testAutoScale
- http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/2021

**Comment by DailyDreaming** [comment]
Also, got rid of one entirely.  It was basically testing: `assert os.path.isdir` which will always be `True`.

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Comment by DailyDreaming** [comment]
Flaky tests:
- src.toil.test.src.fileStoreTest.CachingFileStoreTestWithGoogleJobStore.testCacheEvictionFailCase
- http://jenkins.cgcloud.info/job/toil-pull-requests/3747
- src.toil.test.provisioners.aws.awsProvisionerTest.AWSAutoscaleTestMultipleNodeTypes.testAutoScale
- http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/2044

**Comment by diekhans** [comment]
The reason for not porting require is the generic error messages are less helpful
than more tailored ones.

**Comment by DailyDreaming** [comment]
Regular jenkins did not finish at all and appears to have some issues:
http://jenkins.cgcloud.info/job/toil-pull-requests/3760/console
Failed integration tests:
- src.toil.test.provisioners.gceProvisionerTest.GCEAutoscaleTest.testAutoScale
- src.toil.test.provisioners.gceProvisionerTest.GCEAutoscaleTest.testSpotAutoScale
- http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/2057

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Pull request merged**